File: VIF_git/GSDS-Qual/LeetCode/week4/21-2/suho/2.py
class LinkedNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def split(l):
    node = l
    len = 1

    while(node.next != None): 
        node = node.next
        len += 1

    node = l
    for i in range(len//2 - 1): node = node.next
    right = node.next
    node.next = None

    return l, right


def merge(l1, l2):
    lnode = l1
    rnode = l2

    if lnode.val > rnode.val:
        l = lnode
        lnode = lnode.next
    else:
        l = rnode
        rnode = rnode.next
    l.next = None
    l_head = l

    while lnode != None and rnode != None:
        if lnode.val > rnode.val:
            l.next = lnode
            lnode = lnode.next
            l = l.next
            l.next = None
        else:
            l.next = rnode
            rnode = rnode.next
            l = l.next
            l.next = None

    while lnode != None:
        l.next = lnode
        lnode = lnode.next
        l = l.next
        l.next = None

    while rnode != None:
        l.next = rnode
        rnode = rnode.next
        l = l.next
        l.next = None

    return l_head

def sortingLL(l):
    if  l.next != None:
        left, right = split(l)

        left = sortingLL(left)
        right = sortingLL(right)

        l = merge(left, right)
   
    return l


node = LinkedNode(1)
l = node
node.next = LinkedNode(2)
node = node.next
node.next = LinkedNode(3)
node = node.next
node.next = LinkedNode(4)

l = sortingLL(l)

node = l
while(node != None):
    print(node.val)
    node = node.next

File: VIF_git/GSDS-Qual/LeetCode/week4/21-2/suho/2.cpp
# include <iostream>
using namespace std;

struct LinkedNode{
        int val;
        LinkedNode *next = nullptr;
        LinkedNode(int val) : val(val) {}
};

LinkedNode* split(LinkedNode *l){
    LinkedNode *node = l;
    int len = 1;

    while(node->next != nullptr){
        node = node->next;
        len++;
    }

    node = l;
    for(int i = 0; i <  len/2 - 1; i++) node = node->next;
    LinkedNode *right = node->next;
    node->next = nullptr;

    return right;
}


LinkedNode* merge(LinkedNode *l1, LinkedNode *l2){
    LinkedNode *lnode = l1;
    LinkedNode *rnode = l2;
    LinkedNode *l;

    if(lnode->val > rnode->val){
        l = lnode;
        lnode = lnode->next;
    }
    else{
        l = rnode;
        rnode = rnode->next;
    }
    l->next = nullptr;
    LinkedNode *l_head = l;

    while(lnode != nullptr && rnode != nullptr){
        if(lnode->val > rnode->val){
            l->next = lnode;
            lnode = lnode->next;
            l = l->next;
            l->next = nullptr;
        }
        else{
            l->next = rnode;
            rnode = rnode->next;
            l = l->next;
            l->next = nullptr;
        }
    }
    while(lnode != nullptr){
        l->next = lnode;
        lnode = lnode->next;
        l = l->next;
        l->next = nullptr;
    }
    while(rnode != nullptr){
        l->next = rnode;
        rnode = rnode->next;
        l = l->next;
        l->next = nullptr;
    }

    return l_head;
}

LinkedNode* sortingLL(LinkedNode *l){
    if(l->next != nullptr){
        LinkedNode* right = split(l);

        l = sortingLL(l);
        right = sortingLL(right);

        l = merge(l, right);
    }
   
    return l;
}

int main(){
    LinkedNode node1 = LinkedNode(1);
    LinkedNode node2 = LinkedNode(2);
    LinkedNode node3 = LinkedNode(3);
    LinkedNode node4 = LinkedNode(4);
    LinkedNode node5 = LinkedNode(5);
    
    LinkedNode *lh = &node1;
    LinkedNode *l = lh;
    l->next = &node2;
    l = l->next;
    l->next = &node3;
    l = l->next;
    l->next = &node4;
    l = l->next;
    l->next = &node5;
 
    l = sortingLL(lh);

    while(l != nullptr){
        printf("%d\n",l->val);
        l = l->next;
    }

    return 0;
}
